boot/freeldr/freeldr/arch/i386/pcmem.c

```diff
@@ -549,6 +549,12 @@ PcMemFinalizeMemoryMap(
     ReserveMemory(MemoryMap, STACKLOW, STACKADDR - STACKLOW, LoaderOsloaderStack, "FreeLdr stack");
     ReserveMemory(MemoryMap, FREELDR_BASE, FrLdrImageSize, LoaderLoadedProgram, "FreeLdr image");
 
+    /* We need to make sure that we don't allocate the 4 MB chunk of physical memory starting at
+     * 0x40000000. The logic used in MempAllocatePTE means that the kernel mapping would be at
+     * 0xc0000000, which is where Windows needs the page directory to be mapped to (see
+     * MempAllocatePageTables). */
+    ReserveMemory(MemoryMap, 0x40000000, 0x400000, LoaderReserve, "Reserved");
```

Comment:
`0x400000` is enough for all the page tables, but the page directory itself (at `0xC0400000`) takes another page. I guess that should be included, too...

Reply:
I don't think these NT-specific exclusion memory areas must be hardcoded in the arch-specific pcmem.c but instead this should be done in the NT loader (in one of the files under the ntldr subdirectory).

Reply:
It is architecture-specific, though - it's 4 MB because that's the size of an x86 page directory. If Freeldr used PAE, say, the the values would be different.

Reply:
Yes, that's why it would better go somewhere in ntldr/arch/i386/winldr.c instead.
(The arch/i386/pcmem.c & co files are for hardware-ARC-emulation support only.)